[5.5] Use table aliases when calling a self-referencing HasManyThrough

JayBizzle opened the pull request:
See #21259 for full-details.

This fix has been implemented in a similar way that table aliases are applied when using a self-referencing `BelongsTo` relationship.

I have targeted this at 5.5, but not sure if this is classed as a breaking change as some people may be relying on the current broken behaviour. If so, I can target this at 5.6 no problem.

JayBizzle committed:
Use table aliases when calling a self-referencing HasManyThrough relationship

browner12 committed:
add `notifyNow` method to Notifiables (#21795)

allows user to override notifications with the `ShouldQueue` interface, and send them immediately.

roberto-aguilar committed:
Create assertion to determine if a cookie is expired (#21793)

Given that the `\Illuminate\Cookie\CookieJar::forget` method expires the
given cookie to trigger its deletion, this will allow to test those
scenarios.

This new assertion will

- Fail if the cookie is not present at all.
- Fail if the cookie is present, but it is not expired.
- Pass if the cookie is present and is expired.

jmarcher commented:
Can you please include some tests?

JayBizzle commented:
Based on the functionality of the `BelongsTo` relation, there are no tests

themsaid commented:
@JayBizzle you can write an integration test where you hit an actual DB and make sure it returns results.

JayBizzle commented:
Yes, an integration test should be no problem

tillkruss committed:
update v5.5 changelog

jacobgacosta committed:
Add assertion to verify that a cookie is missing (#21803)

JayBizzle committed:
Add integration test

mattwells committed:
Add Type Hint to Model Factory Stub (#21810)

I find myself adding this type hint to every model factory that I
create to help IDEs auto complete the $factory object.

imbrish committed:
Set current user before firing authenticated event (#21790)

jfadich committed:
allow the distinct validation rule to optionally ignore case (#21757)

conform to style guide

add unicode support

escape slashes in regular expression

4refael committed:
Minor accessibility improvements (#21817)

themsaid committed:
[5.5] fix model refresh with global scopes (#21815)

*        fix model refresh with global scopes

*    fix style

*    fix style

ankitpokhrel committed:
Remove non-existent argument (#21814)

themsaid committed:
fix scheduling a non queueable job (#21820)

taylorotwell committed:
version

themsaid committed:
update release notes

themsaid committed:
update release notes

doomtickle committed:
Fixed typo in comments of Queue/Worker.php

akiyamaSM committed:
Update DatabaseNotificationCollection.php (#21841)

Improve documentation :D

taylorotwell committed:
formatting

ahmed-aliraqi committed:
Fix controller does not exist issue (#21844)

jesseschutt committed:
Adds “kin” to uncountable list (#21843)

Signed-off-by: Jesse Schutt <jesse.schutt@zaengle.com>

GrahamCampbell committed:
Removed uneeded doc

iget-esoares committed:
Fix Resource `collection()` phpdoc return type (#21866)

drfraker committed:
update docblock to include the Throwable (#21878)

* update dockblock to include the Throwable

* Update helpers.php

drfraker committed:
Add possible thrown exception to the docblock of function (#21877)

* Add possible thrown exception to the docblock of function

* Update helpers.php

themsaid commented:
@JayBizzle I think the example you have is wrong, use case doesn't feel right to me, can you explain in plain english the relationship between the user and his referrals?

themsaid commented:
Here's an example self-referencing HasManyThrough use case:

```
<?php

namespace Illuminate\Tests\Integration\Database\EloquentHasManyThroughTest;

use Orchestra\Testbench\TestCase;
use Illuminate\Support\Facades\Schema;
use Illuminate\Database\Eloquent\Model;

/**
 * @group integration
 */
class EloquentHasManyThroughTest extends TestCase
{
    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('app.debug', 'true');

        $app['config']->set('database.default', 'testbench');

        $app['config']->set('database.connections.testbench', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    public function setUp()
    {
        parent::setUp();

        Schema::create('users', function ($table) {
            $table->increments('id');
            $table->integer('team_id')->nullable();
            $table->string('name');
        });

        Schema::create('teams', function ($table) {
            $table->increments('id');
            $table->integer('owner_id');
        });
    }

    /**
     * @test
     */
    public function basic_create_and_retrieve()
    {
        $user = User::create(['name' => str_random()]);

        $team1 = Team::create(['owner_id' => $user->id]);
        $team2 = Team::create(['owner_id' => $user->id]);

        $member1 = User::create(['name' => str_random(), 'team_id' => $team1->id]);
        $member2 = User::create(['name' => str_random(), 'team_id' => $team1->id]);

        $notMember = User::create(['name' => str_random()]);


        // Returns the correct results for a self referencing relationship
        $user->members->toArray();
    }
}


class User extends Model
{
    public $table = 'users';
    public $timestamps = false;
    protected $guarded = ['id'];

    public function members()
    {
        return $this->hasManyThrough(self::class, Team::class, 'owner_id', 'team_id');
    }
}

class Team extends Model
{
    public $table = 'teams';
    public $timestamps = false;
    protected $guarded = ['id'];
}
``

It works fine, so I don't think there's a bug that needs to be fixed.

themsaid committed:
Apply fixes from StyleCI (#21880)

JayBizzle commented:
Give me a second to go through your example....

themsaid commented:
@JayBizzle I've updated the example with better relationship name for clarity.

themsaid committed:
add EloquentHasManyThroughTest

themsaid committed:
naming

JayBizzle commented:
Okay, testing now....

JayBizzle commented:
@themsaid In your example, if you did the following...

```
User::has('teamMates')->get();
```

You would get no results, whereas you should get 1 which you do after applying my fix.

Your example isn't quite what we were trying to achieve, but it still suffers the same problem.

themsaid commented:
@JayBizzle I see, can you rebase your PR with my tests instead of the those your wrote? We can start checking the approach you took and see if there's a simpler one.

JayBizzle commented:
Yes, ill get that done for you.

I guess this PR name is a bit misleading as this issue doesn't affect ALL self-referencing HasManyThrough relationship calls, only ones that use sub-queries.

themsaid commented:
@JayBizzle yeah it got me confused, but it's ok now that we can see the issue.

themsaid commented:
@JayBizzle git went crazy, I suggest that you open a new PR with just the changes in https://raw.githubusercontent.com/JayBizzle/framework/54edc2d20e96e2f429d24c21e33d6a880cd5d044/src/Illuminate/Database/Eloquent/Relations/HasManyThrough.php, you can copy the file in your new PR and that's it.

JayBizzle commented:
Yeah, i totally messed that up :)
